src/misc/matching.c

```diff
@@ -1007,6 +1007,450 @@ static igraph_error_t igraph_i_maximum_bipartite_matching_weighted(
+            VECTOR(*matching)[v] = w;
+            VECTOR(*matching)[w] = v;
+        }
+
+        // last edge of path is {v,w}
+        v = VECTOR(path)[igraph_vector_int_size(&path) - 2];
+        w = VECTOR(path)[igraph_vector_int_size(&path) - 1];
+        VECTOR(*matching)[v] = w;
+        VECTOR(*matching)[w] = v;
+
+        *matching_size = *matching_size + 1;
+
+#ifdef MATCHING_DEBUG
+        debug("Matching:");
+        igraph_vector_int_print(matching);
+#endif
+
+        igraph_i_maximum_matching_find_aug_path(graph, matching, &path);
+#ifdef MATCHING_DEBUG
+        debug("Path:");
+        igraph_vector_int_print(&path);
+#endif
+    }
+#ifdef MATCHING_DEBUG
+    debug("Matching:");
+    igraph_vector_int_print(matching);
+#endif
+
+    igraph_vector_int_destroy(&path);
+    IGRAPH_FINALLY_CLEAN(1);
+
+    return IGRAPH_SUCCESS;
+}
+
+igraph_error_t igraph_i_maximum_matching_find_aug_path(const igraph_t *graph, igraph_vector_int_t *matching, igraph_vector_int_t *path) {
+    igraph_integer_t n = igraph_vcount(graph);
+    igraph_integer_t v, v_contraction, w, w_contraction, x, blossom_root, current;
+    igraph_bool_t finished, path_found, contracted;
+
+    // Forest F built using set of vectors
+    igraph_vector_int_t f_parent;
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&f_parent, n);
+    igraph_vector_bool_t f_is_root;
+    IGRAPH_CHECK(igraph_vector_bool_init(&f_is_root, n));
+    IGRAPH_FINALLY(igraph_vector_bool_destroy, &f_is_root);
+
+    // root vector updated whenever vertex is added to F
+    igraph_vector_int_t root;
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&root, n);
+
+    // distance from root vector updated whenever vertex is added to F
+    igraph_vector_int_t distance;
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&distance, n);
+
+    // contraction vector updated with blossoms saying what vertex a vertex has been contracted to
+    igraph_vector_int_t contraction;
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&contraction, n);
+    // set each vector to be contracted to itself to start
+    for (igraph_integer_t i = 0; i < n; i++) {
+        VECTOR(contraction)[i] = i;
+    }
+
+    igraph_vector_int_t neighbors;
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&neighbors, n);
+
+    // queue of vertices with distance(v, root(v)) even
+    igraph_dqueue_int_t queue;
+    IGRAPH_CHECK(igraph_dqueue_int_init(&queue, n));
+    IGRAPH_FINALLY(igraph_dqueue_int_destroy, &queue);
+
+    // vertex mark vector
+    igraph_vector_bool_t v_mark;
+    IGRAPH_CHECK(igraph_vector_bool_init(&v_mark, n));
+    IGRAPH_FINALLY(igraph_vector_bool_destroy, &v_mark);
+
+    // edge mark matrix
+    igraph_matrix_bool_t e_mark;
+    IGRAPH_CHECK(igraph_matrix_bool_init(&e_mark, n, n));
+    IGRAPH_FINALLY(igraph_matrix_bool_destroy, &e_mark);
+
+    // blossom mark vector, used to find the root of a blossom
+    igraph_vector_bool_t b_mark;
+    IGRAPH_CHECK(igraph_vector_bool_init(&b_mark, n));
+    IGRAPH_FINALLY(igraph_vector_bool_destroy, &b_mark);
```

Comment:
Such pairs can be replaced with the more concise:

```suggestion
    IGRAPH_VECTOR_BOOL_INIT_FINALLY(&b_mark, n);
```

Boolean vector that are not resized can be replaced with a more space-efficient `igraph_bitset_t`. This is a minor nitpick that is better fixed only when everything else is addressed.